File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/brick/BaseBrickTest.java
/*
 * Copyright © 2017-2020 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit.brick;

import android.view.View;

import com.wayfair.brickkit.BrickDataManager;
import com.wayfair.brickkit.viewholder.BrickViewHolder;
import com.wayfair.brickkit.padding.BrickPadding;
import com.wayfair.brickkit.size.BrickSize;

import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.runner.AndroidJUnit4;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public class BaseBrickTest {
    private static final int PADDING = 3;
    private final BrickSize brickSize = mock(BrickSize.class);
    private final BrickDataManager brickDataManager = mock(BrickDataManager.class);
    private BaseBrick brick = new TestBaseBrick(brickSize);

    @Test
    public void testBrickSizeBrickPaddingConstructor() {
        BrickPadding padding = mock(BrickPadding.class);
        when(padding.getInnerBottomPadding()).thenReturn(PADDING);

        TestBaseBrick brick = new TestBaseBrick(brickSize, padding);

        assertEquals(brickSize, brick.getSpanSize());
        assertEquals(PADDING, brick.getPadding().getInnerBottomPadding());
    }

    @Test
    public void testHidden() {
        assertFalse(brick.isHidden());

        brick.setHidden(true);
        assertTrue(brick.isHidden());
    }

    @Test
    public void testInFirstRow() {
        assertFalse(brick.isInFirstRow());

        brick.setInFirstRow(true);
        assertTrue(brick.isInFirstRow());
    }

    @Test
    public void testInLastRow() {
        assertFalse(brick.isInLastRow());

        brick.setInLastRow(true);
        assertTrue(brick.isInLastRow());
    }

    @Test
    public void testOnLeftWall() {
        assertFalse(brick.isOnLeftWall());

        brick.setOnLeftWall(true);
        assertTrue(brick.isOnLeftWall());
    }

    @Test
    public void testOnRightWall() {
        assertFalse(brick.isOnRightWall());

        brick.setOnRightWall(true);
        assertTrue(brick.isOnRightWall());
    }

    @Test
    public void testRefreshItem() {
        brick.refreshItem();

        brick.setDataManager(brickDataManager);

        brick.refreshItem();
        verify(brickDataManager).refreshItem(brick);
    }

    @Test
    public void testSmoothScroll() {
        brick.smoothScroll();

        brick.setDataManager(brickDataManager);

        brick.smoothScroll();
        verify(brickDataManager).smoothScrollToBrick(brick);
    }

    @Test
    public void testAddLastTo() {
        brick.addLastTo(brickDataManager);

        verify(brickDataManager).addLast(brick);
    }

    @Test
    public void testAddFirstTo() {
        brick.addFirstTo(brickDataManager);

        verify(brickDataManager).addFirst(brick);
    }

    private static final class TestBaseBrick extends BaseBrick {

        private TestBaseBrick(BrickSize spanSize, BrickPadding padding) {
            super(spanSize, padding);
        }

        private TestBaseBrick(BrickSize spanSize) {
            super(spanSize);
        }

        @Override
        public void onBindData(BrickViewHolder holder) { }

        @Override
        public int getLayout() {
            return 0;
        }

        @Override
        public BrickViewHolder createViewHolder(View itemView) {
            return null;
        }
    }
}

File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/brick/BaseBrickTest.kt
/*
 * Copyright © 2017-2021 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit.brick

import android.view.View
import com.wayfair.brickkit.BrickDataManager
import com.wayfair.brickkit.viewholder.BrickViewHolder
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions

class BaseBrickTest {
    private val brickDataManager: BrickDataManager = mock()

    private val brick: BaseBrick = object : BaseBrick(mock(), mock()) {
        override fun onBindData(holder: BrickViewHolder) {}
        override val layout = LAYOUT
        override fun createViewHolder(itemView: View): BrickViewHolder = mock()
    }

    @Test
    fun testHidden() {
        brick.setDataManager(brickDataManager)
        brick.setDataManager(null)

        assertFalse(brick.isHidden)

        brick.isHidden = true
        assertTrue(brick.isHidden)

        brick.isHidden = false
        assertFalse(brick.isHidden)

        verifyNoInteractions(brickDataManager)
    }

    @Test
    fun testHidden_falseToTrue() {
        brick.isHidden = false
        brick.setDataManager(brickDataManager)

        brick.isHidden = true
        assertTrue(brick.isHidden)
        verify(brickDataManager).hideItem(brick)
    }

    @Test
    fun testHidden_falseToFalse() {
        brick.isHidden = false
        brick.setDataManager(brickDataManager)

        brick.isHidden = true
        assertTrue(brick.isHidden)
        verify(brickDataManager, never()).showItem(brick)
    }

    @Test
    fun testTag_nullDataManager() {
        assertNull(brick.tag)

        brick.tag = TAG_1

        brick.setDataManager(brickDataManager)

        brick.tag = TAG_1

        verifyNoInteractions(brickDataManager)
    }

    @Test
    fun testLayout() {
        assertEquals(LAYOUT, brick.layout)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun testPlaceholderLayout() {
        brick.placeholderLayout
    }

    @Test
    fun testRefreshItem() {
        brick.refreshItem()
        brick.setDataManager(brickDataManager)
        brick.refreshItem()

        verify(brickDataManager).refreshItem(brick)
    }

    @Test
    fun testAddLastTo() {
        brick.addLastTo(brickDataManager)

        verify(brickDataManager).addLast(brick)
    }

    @Test
    fun testAddFirstTo() {
        brick.addFirstTo(brickDataManager)

        verify(brickDataManager).addFirst(brick)
    }

    @Test
    fun testIsDataReady() {
        assertTrue(brick.isDataReady)
    }

    @Test
    fun testSmoothScroll() {
        brick.setDataManager(brickDataManager)
        brick.smoothScroll()

        verify(brickDataManager).smoothScrollToBrick(brick)
    }

    @Test
    fun testSmoothScroll_nullDataManager() {
        brick.setDataManager(brickDataManager)
        brick.setDataManager(null)

        brick.smoothScroll()

        verifyNoInteractions(brickDataManager)
    }

    companion object {
        private const val LAYOUT = 1234
        private const val TAG_1 = "tag 1"
    }
}
